### PR TITLE
Support occ based on sequence numbers.

### PR DESCRIPTION
**Pull request by andreidan**
Code migrated from https://github.com/elastic/elasticsearch/commit/f6b5d7e01387c09a449f67e2678cf852f58770b1

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: es/es-server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 1]
```diff
@@ -1178,24 +1178,22 @@ public long version() {
...
-        }
-
         @Override
         public Source getSource() {
             return new Source(source, routing);
         }
 
         private void write(final StreamOutput out) throws IOException {
-            out.writeVInt(SERIALIZATION_FORMAT);
+            final int format = out.getVersion().onOrAfter(Version.V_4_0_0) ? SERIALIZATION_FORMAT : FORMAT_6_0;
```
Comment (andreidan) [inline es/es-server/src/main/java/org/elasticsearch/index/translog/Translog.java]: to confirm - we use our own version as the serialization format?

File: blackbox/docs/general/occ.rst [hunk 2]
```diff
@@ -34,62 +33,75 @@ on every update.
...
 
     cr> update locations set description = 'Updated description'
-    ... where id=5 and "_version" = 2;
+    ... where id=5 and "_seq_no" = 222 and "_primary_term" = 1;
     UPDATE OK, 0 rows affected (... sec)
 
 Optimistic Delete
 =================
 
 Of course the same can be done when deleting a row::
```
Comment (mfussenegger) [inline blackbox/docs/general/occ.rst]:
```suggestion
The same can be done when deleting a row::
```

Not related to the PR, but `Of course` implies that something is trivial or obvious, which for someone reading the docs to learn something is almost never the case.

File: es/es-server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 3]
```diff
@@ -641,19 +643,21 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
...
 
-        return index(getEngine(), operation);
+        return index(engine, operation);
     }
 
     public static Engine.Index prepareIndex(DocumentMapperForType docMapper, Version indexCreatedVersion, SourceToParse source, long seqNo,
-            long primaryTerm, long version, VersionType versionType, Engine.Operation.Origin origin, long autoGeneratedIdTimestamp,
-            boolean isRetry) {
+                                            long primaryTerm, long version, VersionType versionType, Engine.Operation.Origin origin,
+                                            long autoGeneratedIdTimestamp, boolean isRetry,
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I think we should adopt our formatting style (either all args on 1 line, or 1 arg per line) when we take over code. 

(Fine for me to ignore this for this PR, but for the future)

File: es/es-server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 4]
```diff
@@ -2100,6 +2114,21 @@ private static void persistMetadata(
         }
     }
 
+    /**
+     * If an index/update/get/delete operation is using the special `_doc` type, then we replace
+     * it with the actual type that is being used in the mappings so that users may use typeless
+     * APIs with indices that have types.
+     */
+    private String resolveType(String type) {
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Do we need this? We always have only 1 type.

Maybe out of scope for this PR, but I think we could go through some of the type usages and remove them

File: es/es-server/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 1 above)
Comment (mfussenegger) [reply es/es-server/src/main/java/org/elasticsearch/index/translog/Translog.java]: Yep, there is no other Version anymore.

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+import static io.crate.testing.TestingHelpers.printedTable;
+import static org.hamcrest.core.Is.is;
+
+public class SeqNoBasedOCCIntegrationTest extends SQLTransportIntegrationTest {
+
+    @Test
+    public void testDeleteWhereSeqNoAndTerm() throws Exception {
+        execute("create table t (x integer primary key, y string) with (number_of_replicas=0)");
+        ensureYellow();
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]:
```suggestion
```

`ensureYellow` is not needed due to our `write.wait_for_active_shards` default.

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        assertEquals(1L, response.rowCount());
+        Long seqNo = (Long) response.rows()[0][0];
+        Long primaryTerm = (Long) response.rows()[0][1];
+
+        execute("delete from t where x = 1 and _seq_no = ? and _primary_term = ?",
+            new Object[]{seqNo, primaryTerm});
+        assertEquals(1L, response.rowCount());
+
+        // Validate that the row is really deleted
+        refresh();
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]:
```suggestion
```
`refresh` shouldn't be necessary with the primary key lookup below.

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        execute("select * from t where x = 1");
+        assertEquals(0, response.rowCount());
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]: Could move the comment above into the assertion as message:
```suggestion
        assertThat("Records with x=1 must be deleted", response.rowCount(), is(0L));
```

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+import io.crate.exceptions.VersioninigValidationException;
+import org.junit.Test;
+
+import static io.crate.testing.TestingHelpers.printedTable;
+import static org.hamcrest.core.Is.is;
+
+public class SeqNoBasedOCCIntegrationTest extends SQLTransportIntegrationTest {
+
+    @Test
+    public void testDeleteWhereSeqNoAndTerm() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]:
```suggestion
    public void testDeleteWhereSeqNoAndTermThatMatch() throws Exception {
```

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        assertEquals(1L, response.rowCount());
+
+        // Validate that the row is really deleted
+        refresh();
+        execute("select * from t where x = 1");
+        assertEquals(0, response.rowCount());
+    }
+
+    @Test
+    public void testDeleteWithConflict() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]:
```suggestion
    public void testDeleteWithSeqNoAndTermThatConflict() throws Exception {
```

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        execute("create table t (x integer primary key, y string)");
+        ensureYellow();
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]:
```suggestion
```

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+        execute("delete from t where x = 1 and _seq_no = ? and _primary_term = ?",
+                new Object[]{preUpdateSeqNo, preUpdatePrimaryTerm});
+        assertEquals(0, response.rowCount());
+    }
+
+    @Test
+    public void testUpdateWhereSeqNoAndPrimaryTermWithPrimaryKey() throws Exception {
+        execute("create table t (x integer primary key, y string)");
+        ensureYellow();
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]:
```suggestion
```

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        assertEquals(1L, response.rowCount());
+        Long preUpdateSeqNo = (Long) response.rows()[0][0];
+        Long preUpdatePrimaryTerm = (Long) response.rows()[0][1];
+
+        execute("update t set y = ? where x = ? and _seq_no = ? and _primary_term = ?",
+                new Object[]{"ok now panic", 1, preUpdateSeqNo, preUpdatePrimaryTerm});
+        assertEquals(1L, response.rowCount());
+
+        // Validate that the row is really updated
+        refresh();
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]:
```suggestion
```

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        execute("select y from t where x = 1");
+        assertEquals(1L, response.rowCount());
+        assertEquals("ok now panic", response.rows()[0][0]);
+    }
+
+    @Test
+    public void testUpdateWhereSeqNoAndPrimaryTermWithoutPrimaryKey() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]: Can we turn this into a unit test?

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+
+        // Validate that the row is really NOT updated
+        refresh();
+        execute("select y from t where x = 1");
+        assertEquals(1L, response.rowCount());
+        assertEquals("don't panic", response.rows()[0][0]);
+    }
+
+    @Test
+    public void testSelectWhereSeqNoAndPTWithoutPrimaryKey() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]: Can we turn this into a unit test ? (several hundred ms, vs a couple of ms runtime)

File: sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+        execute("create table t (x integer primary key, y string)");
+        ensureYellow();
+        expectedException.expect(SQLActionException.class);
+        expectedException.expectMessage(VersioninigValidationException.SEQ_NO_AND_PRIMARY_TERM_USAGE_MSG);
+        execute("select _seq_no, _primary_term from t where y = 'hello' and _seq_no = 1 and _primary_term = 2");
+    }
+
+    @Test
+    public void testSelectGroupBySeqNoPrimaryTerm() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]: What's special about this test case?

Comment (andreidan) [reply sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]: I wondered that myself. I migrated some of the tests from https://github.com/crate/crate/blob/master/sql/src/test/java/io/crate/integrationtests/VersionHandlingIntegrationTest.java#L174 

I'm not sure what code path was meant to be covered by the grouping case to be honest. Happy to drop it if you think it's not worth it.

Comment (mfussenegger) [reply sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java]: yep I'd drop it. Maybe this is legacy from a time when our planner was less modular/compositional

**Pull request merged**